File: script/course_completion.py
import json
import sys
import requests

HEADERS = {
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36',
    'x-coursera-application': 'ondemand',
    'x-coursera-version': 'cde5f24972aff1ebd6447e911113e781b9c52f7f',
    'x-requested-with': 'XMLHttpRequest',
}

BASE_URL = 'https://www.coursera.org/api/'


class Skipera:
    def __init__(self, course_slug, cauth_token, csrf_token):
        self.moduleCount = 0
        self.user_id = None
        self.course_id = None
        self.base_url = BASE_URL
        self.session = requests.Session()
        self.session.headers.update(HEADERS)
        self.session.cookies.update({
            'CAUTH': cauth_token,
            'CSRF3-Token': csrf_token
        })
        self.course = course_slug
        if self.get_userid() == 0:
            sys.exit(1)

    def get_userid(self):
        r = self.session.get(self.base_url + "adminUserPermissions.v1?q=my").json()
        try:
            self.user_id = r["elements"][0]["id"]
        except KeyError:
            return 0
        return 1

    def get_modules(self):
        r = self.session.get(self.base_url + f"onDemandCourseMaterials.v2/?q=slug&slug={self.course}&includes=modules").json()
        self.course_id = r["elements"][0]["id"]
        modules = r.get("linked", {}).get("onDemandCourseMaterialModules.v1", [])


    def get_items(self):
        r = self.session.get(self.base_url + "onDemandCourseMaterials.v2/", params={
            "q": "slug",
            "slug": self.course,
            "includes": "passableItemGroups,passableItemGroupChoices,items,tracks,gradePolicy,gradingParameters",
            "fields": "onDemandCourseMaterialItems.v2(name,slug,timeCommitment,trackId)",
            "showLockedItems": "true"
        }).json()
        items = r.get("linked", {}).get("onDemandCourseMaterialItems.v2", [])
        for video in items:
            self.moduleCount += 1
            self.watch_item(video["id"])

    def watch_item(self, item_id):
        r = self.session.post(
            self.base_url + f"opencourse.v1/user/{self.user_id}/course/{self.course}/item/{item_id}/lecture/videoEvents/ended?autoEnroll=false",
            json={"contentRequestBody": {}}).json()

        if r.get("contentResponseBody") is None:
            self.read_item(item_id)

    def read_item(self, item_id):
        r = self.session.post(self.base_url + "onDemandSupplementCompletions.v1", json={
            "courseId": self.course_id,
            "itemId": item_id,
            "userId": int(self.user_id)
        })
        


def main():
    if len(sys.argv) < 4:
        print("Usage: python course_completion.py <course_slug> <CAUTH> <CSRF3-Token>")
        return

    course_slug = sys.argv[1]
    cauth_token = sys.argv[2]
    csrf_token = sys.argv[3]

    skipera = Skipera(course_slug, cauth_token, csrf_token)
    skipera.get_modules()
    skipera.get_items()
    print(skipera.moduleCount)


if __name__ == '__main__':
    main()
    # print(1)

File: lib/courseCompletion.js
const BASE_URL = 'https://www.coursera.org/api/';

const DEFAULT_HEADERS = {
  'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36',
  'x-coursera-application': 'ondemand',
  'x-coursera-version': 'cde5f24972aff1ebd6447e911113e781b9c52f7f',
  'x-requested-with': 'XMLHttpRequest'
};

async function fetchJson(url, options) {
  const resp = await fetch(url, options);
  const text = await resp.text();
  let json;
  try { json = text ? JSON.parse(text) : {}; } catch (_) { json = {}; }
  if (!resp.ok) {
    const err = new Error(`HTTP ${resp.status} for ${url}`);
    err.status = resp.status;
    err.body = json;
    throw err;
  }
  return json;
}

function buildCookie(cAuth, csrf) {
  const parts = [];
  if (cAuth) parts.push(`CAUTH=${cAuth}`);
  if (csrf) parts.push(`CSRF3-Token=${csrf}`);
  return parts.join('; ');
}

async function getUserId(headers) {
  const data = await fetchJson(`${BASE_URL}adminUserPermissions.v1?q=my`, { headers });
  try {
    return data.elements[0].id;
  } catch (_) {
    throw new Error('Unable to resolve user id');
  }
}

async function getCourseIdAndModules(courseSlug, headers) {
  const url = `${BASE_URL}onDemandCourseMaterials.v2/?q=slug&slug=${encodeURIComponent(courseSlug)}&includes=modules`;
  const data = await fetchJson(url, { headers });
  const courseId = data.elements?.[0]?.id;
  if (!courseId) throw new Error('Unable to resolve course id');
  return { courseId };
}

async function getItems(courseSlug, headers) {
  const params = new URLSearchParams({
    q: 'slug',
    slug: courseSlug,
    includes: 'passableItemGroups,passableItemGroupChoices,items,tracks,gradePolicy,gradingParameters',
    fields: 'onDemandCourseMaterialItems.v2(name,slug,timeCommitment,trackId)',
    showLockedItems: 'true'
  });
  const url = `${BASE_URL}onDemandCourseMaterials.v2/?${params.toString()}`;
  const data = await fetchJson(url, { headers });
  return data.linked?.['onDemandCourseMaterialItems.v2'] || [];
}

async function postVideoEnded(userId, courseSlug, itemId, headers) {
  const url = `${BASE_URL}opencourse.v1/user/${userId}/course/${courseSlug}/item/${itemId}/lecture/videoEvents/ended?autoEnroll=false`;
  return await fetchJson(url, { method: 'POST', headers: { ...headers, 'content-type': 'application/json' }, body: JSON.stringify({ contentRequestBody: {} }) });
}

async function postSupplementCompletion(courseId, itemId, userId, headers) {
  const url = `${BASE_URL}onDemandSupplementCompletions.v1`;
  return await fetchJson(url, { method: 'POST', headers: { ...headers, 'content-type': 'application/json' }, body: JSON.stringify({ courseId, itemId, userId: Number(userId) }) });
}

async function runCourseCompletion(courseSlug, cAuth, csrf) {
  const headers = { ...DEFAULT_HEADERS, cookie: buildCookie(cAuth, csrf) };
  const userId = await getUserId(headers);
  const { courseId } = await getCourseIdAndModules(courseSlug, headers);
  const items = await getItems(courseSlug, headers);

  let moduleCount = 0;
  for (const item of items) {
    const itemId = item.id;
    if (!itemId) continue;
    moduleCount += 1;
    const r = await postVideoEnded(userId, courseSlug, itemId, headers).catch(() => ({}));
    if (!r || r.contentResponseBody == null) {
      await postSupplementCompletion(courseId, itemId, userId, headers).catch(() => {});
    }
  }

  return moduleCount;
}

module.exports = { runCourseCompletion };


